### PR TITLE
stocks-portfolio example

### PR DESCRIPTION
**Pull request by xDiesel76**
# Description

New example on making a stocks-portfolio using YFinance module.

## Status

**READY**

## Type of change

- Documentation (fix or adds documentation)

# How Has This Been Tested?

It was read through by me and people in Matrix channel.

# Checklist:

- [Y] I have performed a self-review of my own code
- [Y] I have made corresponding changes to the documentation (if applicable)
- [Y] I have added tests that prove my fix is effective or that my feature works
- [Y] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: docs/examples/stocks-portfolio.md [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+from opsdroid.matchers import match_regex
+from opsdroid.skill import Skill
+
+conn = sqlite3.connect('/path/to/my/database')
+c = conn.cursor()
+```
+
+### Recieving Input
+
+Now we have to collect what the bot will revieve, we will be doing this by first creating a class that inherits the Skill class. Inside we will be using the special opsdroid function which is `@match_regex`. This function will send the message, and to collect the input we will be using `(.*)` beside the message, like this `@match_regex(r"Stock: (.*)")`. We will then be accessing it with `message.regex.group(1)`. We can now assign what we recieve from the bot.
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:60]:
```suggestion
Now we have to collect what the bot will receive, we will be doing this by first creating a class that inherits the Skill class. Inside we will be using the special opsdroid function which is `@match_regex`. This function will send the message, and to collect the input we will be using `(.*)` beside the message, like this `@match_regex(r"Stock: (.*)")`. We will then be accessing it with `message.regex.group(1)`. We can now assign what we receive from the bot. 
```

File: docs/examples/stocks-portfolio.md [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+
+```python
+class StockSkill(Skill):
+
+    @match_regex(r"Stock: (.*)")
+    async def search_info_stock(self, message):
+        stock_name = message.regex.group(1)
+        stock = yf.Ticker(stock_name) 
+```
+Now `stock` is what we recieve from YFinance when we input a stock's symbol.
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:86]:
```suggestion
Now `stock` is what we receive from YFinance when we input a stock's symbol.
```

File: docs/examples/stocks-portfolio.md [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+
+### Using the Data
+
+Now after getting the data and assigning it to a variable we can start using it! We can now return the info to the connector. We will be using the info property but if you are interested on the different properties you can use their [documentation](https://github.com/ranaroussi/yfinance.
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:90]:
```suggestion
After getting the data and assigning it to a variable, we can start using it! We can now return the info to the connector. We will be using the info property, but if you are interested in the different properties, you can use their [documentation](https://github.com/ranaroussi/yfinance).
```

File: docs/examples/stocks-portfolio.md [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+            await message.respond('No Stocks Added, Please Use Command "Add Stock:"')
+            
+        for row in rows:
+            stock = yf.Ticker(row[0])
+            await message.respond(f"Name: {stock.info['longName']}")
+            await message.respond(f"Price: {stock.info['regularMarketOpen']} {stock.info['currency']}")
+            await message.respond(f"About: {stock.info['longBusinessSummary']}")
+```
+
+Now you can add, list and find specific info about stocks. Congratulations! Goodluck with your opsdroid journey! Here is what the final code should look like:
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:339]:
```suggestion
Now you can add, list and find specific info about stocks. Congratulations! Good luck with your opsdroid journey! Here is what the final code should look like:
```

File: docs/examples/stocks-portfolio.md [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+skills:
+  stock:
+    path: /path/to/my/stocks-portfolio
+```
+
+### Imports and Classes Needed
+
+Now that our skill has been configured in the `configuration.yaml` file, we will continue by creating the `__init__.py` file inside of our stocks-portfolio folder and start working on the skill!
+
+The first thing we need to do is to import the `Skill` class and the `regex_matcher` from opsdroid. Then we will need to import `yfinance` and `sqlite3` modules. Your `__init__.py` file should look like this:
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:33]:
```suggestion
The first thing we need to do is to import the `Skill` class and the `regex_matcher` from opsdroid. Then we will need to import `yfinance`, we will also use [opsdroid's memory](https://docs.opsdroid.dev/en/stable/skills/memory.html) to save data into the SQL database.. Your `__init__.py` file should look like this:
```

File: docs/examples/stocks-portfolio.md [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+
+```python
+import yfinance as yf
+import sqlite3
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:37]:
```suggestion

```

File: docs/examples/stocks-portfolio.md [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+Our next step is to connect our `__init__.py` file to a SQLite database. Your `__init__.py` file should look something like this now:
+
+```python
+import yfinance as yf
+import sqlite3
+
+from opsdroid.matchers import match_regex
+from opsdroid.skill import Skill
+
+conn = sqlite3.connect('/path/to/my/database')
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:54]: Feel free to ignore the database suggestions, if you want to use the opsdroid memory you can use `self.opsdroid.memory.put(<key>, <object>)
```suggestion

```

File: docs/examples/stocks-portfolio.md [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+c = conn.cursor()
+```
+
+### Recieving Input
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:58]:
```suggestion
### Receiving Input
```

File: docs/examples/stocks-portfolio.md [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+We want the `['longName']`, `['regularMarketOpen']` and `['longBusinessSummary']`. Now we want the bot to send the data, so we will be using `await message.respond()`. Here is what it should look like:
+
+```python
+class StockSkill(Skill):
+
+    @match_regex(r"Stock: (.*)")
+    async def search_info_stock(self, message):
+        stock_name = message.regex.group(1)
+        stock = yf.Ticker(stock_name)
+        await message.respond(f"Name: {stock.info['longName']}")
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:233]: I'm happy to have it send more than one message, just as reference you could send a single message and make the string look like:

```python
await message.respond(f"Name: {stock.info['longName']}\nCurrent Price: {stock.info['regularMarketOpen']} {stock.info['currency']}\nAbout: {stock.info['longBusinessSummary']}")
```

File: docs/examples/stocks-portfolio.md [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+        await message.respond(f"Current Price: {stock.info['regularMarketOpen']} {stock.info['currency']}")
+        await message.respond(f"About: {stock.info['longBusinessSummary']}")
+```
+
+Congratulations! You have now made a bot that can tell you the name, price and info on a stock. You can do this with `Stock: <symbol-of-stock>`. With the basic part of the bot finished we can make the portfolio/saving section.
+
+### Saving and Adding to the Portfolio
+
+#### Creating the Table
```
Comment (FabioRosado) [inline docs/examples/stocks-portfolio.md:242]: Like I said before, i'm happy to have the SQL steps here -  if you want to replace it you could change this step like:

### Putting in opsdroid memory

Opsdroid has a memory class which can be used to persist data, to use opsdroid memory you need to add the following to your configuration:

```yaml
databases:
  sqlite:
```

Then we can add to the database using `await self.opsdroid.memory.put(stocks, params)`

---

You might need to think about how the data is going to be saved/returned

**Review by FabioRosado — CHANGES_REQUESTED** [review]
Hello @xDiesel76  thank you for submitting this PR it looks great, I've added a few suggestions just to fix some minor typos. 

I've also tried to give you a hand with the opsdroid memory if you decide to use it, but it's up to you if you want to use the opsdroid memory or not 😄 👍 

If you don't want to rewrite the whole thing to use opsdroid memory as soon s the typo suggestions are fixed (you can accept the changes with a click of a button), I'm happy to merge it.

File: docs/examples/stocks-portfolio.md (hunk 5 above)
Comment (xDiesel76) [reply docs/examples/stocks-portfolio.md:33]: Hello Fabio, I will try to understand the memory thing more before I rewrite the docs. I think maybe a week.

File: docs/examples/stocks-portfolio.md (hunk 9 above)
Comment (xDiesel76) [reply docs/examples/stocks-portfolio.md:233]: Thank you, I will be adding this

File: docs/examples/stocks-portfolio.md (hunk 10 above)
Comment (xDiesel76) [reply docs/examples/stocks-portfolio.md:242]: Thank you this info really helps!

**Comment by xDiesel76** [comment]
Fixed typos, thanks!

**Review by FabioRosado — APPROVED** [review]
I'm assuming that this PR is ready for a review, thank you for working on this and updating the PR. 
I can see that you merged my suggestions as well, so I'm happy to get this merged 😄 👍

**Pull request merged**